lib/src/Uncertainty/Distribution/GeneralizedExtremeValueFactory.cxx

```diff
@@ -82,6 +94,899 @@ GeneralizedExtremeValue GeneralizedExtremeValueFactory::buildAsGeneralizedExtrem
+      {
+        const Scalar c1 = 1.0 + xi * yi;
+        if (c1 <= SpecFunc::Precision) // can be slightly off
+          return Point(1, -SpecFunc::MaxScalar);
+        ll += -(1.0 + 1.0 / xi) * std::log(c1) - std::pow(c1, -1.0 / xi);
+      }
+    }
+    return Point(1, ll);
+  }
+
+private:
+  Sample sample_;
+};
+
+
+static Scalar GeneralizedExtremeValueFactoryPWM(const Sample & sample, UnsignedInteger r)
+{
+  Scalar s = 0.0;
+  const UnsignedInteger size = sample.getSize();
+  for (UnsignedInteger i = r; i < size; ++ i)
+    s += std::exp(SpecFunc::LogGamma(1.0 * (i + 1)) - SpecFunc::LogGamma(1.0 * (i + 1 - r))) * sample(i, 0);
+  return s / std::exp(SpecFunc::LogGamma(size + 1.0) - SpecFunc::LogGamma(1.0 * (size - r)));
+}
+
+
+DistributionFactoryLikelihoodResult GeneralizedExtremeValueFactory::buildMethodOfLikelihoodMaximizationEstimator(const Sample & sample) const
+{
+  if (sample.getSize() < 3)
+    throw InvalidArgumentException(HERE) << "Error: cannot build a GeneralizedExtremeValue distribution from a sample of size < 3";
+  if (sample.getDimension() != 1)
+    throw InvalidArgumentException(HERE) << "Error: can build a GeneralizedExtremeValue distribution only from a sample of dimension 1, here dimension=" << sample.getDimension();
+
+  const Function objective(new GeneralizedExtremeValueLikelihoodEvaluation(sample));
+  OptimizationProblem problem(objective);
+  problem.setMinimization(false);
+
+  const Scalar zmin = sample.getMin()[0];
+  const Scalar zmax = sample.getMax()[0];
+  const Scalar mean = sample.computeMean()[0];
+
+  // sigma > 0
+  const Point lowerBound({-SpecFunc::MaxScalar, 0.0, -SpecFunc::MaxScalar});
+  const Point upperBound(3, SpecFunc::MaxScalar);
+  const Interval::BoolCollection finiteLowerBound({false, true, false});
+  const Interval::BoolCollection finiteUpperBound(3, false);
+  problem.setBounds(Interval(lowerBound, upperBound, finiteLowerBound, finiteUpperBound));  
+
+  // 1+xi(zi-mu)/sigma > 0
+  Description formulas(2);
+  formulas[0] = OSS() << "sigma + xi * (" << zmax << " - mu)";
+  formulas[1] = OSS() << "sigma + xi * (" << zmin << " - mu)";
+  const SymbolicFunction constraint(Description({"mu", "sigma", "xi"}), formulas);
+  problem.setInequalityConstraint(constraint);
+
+  // pwm for the starting point, see fit.gev function from R mev package
+  const Sample sorted(sample.sort());
+  const Scalar bpwm1 = GeneralizedExtremeValueFactoryPWM(sorted, 1);
+  const Scalar bpwm2 = GeneralizedExtremeValueFactoryPWM(sorted, 2);
+  const Scalar kst = (2.0 * bpwm1 - mean) / (3.0 * bpwm2 - mean) - std::log(2.0) / std::log(3.0);
+  const Scalar xi0 = -(7.859 + 2.9554 * kst) * kst;
+  const Scalar gamma1mXi0 = SpecFunc::Gamma(1.0 - xi0);
+  const Scalar sigma0 = -(2.0 * bpwm1 - mean) * xi0 / (gamma1mXi0 * (1.0 - std::pow(2.0, xi0)));
+  const Scalar mu0 = mean - sigma0 * (gamma1mXi0 - 1.0) / xi0;
+  const Point x0({mu0, sigma0, xi0});
+
+  // solve optimization problem
+  Cobyla solver(problem);
```

Comment:
Yet another hard coded use of Cobyla ;-). We should definitely think about interfacing PRIMA (see #2249)

Reply:
yes, hopefully there will be a C version someday